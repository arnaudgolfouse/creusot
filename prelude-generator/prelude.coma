module Any
  function any_l (a :'a) : 'b
end

module MutBorrow
  type t 'a = { current : 'a; final : 'a; id : int }

  let borrow_final < 'a > (a : 'a) (id : int) (ret (result : t 'a)) =
    any [ k (fin : 'a) -> (! ret { { current = a; final = fin; id = id } })]

  let borrow_mut < 'a > (a : 'a) (ret (result : t 'a)) =
    any [ k (fin : 'a) (id: int) -> (! ret { { current = a; final = fin; id = id } })]

  function borrow_logic (cur fin : 'a) (id : int) : t 'a = { current = cur; final = fin; id = id }

  function get_id (x : t 'a) : int = x.id
  function inherit_id (old_id inherit_path: int) : int
end

module Bool
  use bool.Bool

  function to_int [@inline:trivial] (b : bool) : int = if b then 1 else 0

  function bw_and [@inline:trivial] (a:bool) (b:bool) : bool = a /\ b
  function bw_or [@inline:trivial] (a:bool) (b:bool) : bool = a \/ b
  function bw_xor [@inline:trivial] (a:bool) (b:bool) : bool = not a = b

  predicate le [@inline:trivial] (a:bool) (b:bool) = (not a) \/ b
  predicate lt [@inline:trivial] (a:bool) (b:bool) = (not a) /\ b
  predicate ge [@inline:trivial] (a:bool) (b:bool) = a \/ not b
  predicate gt [@inline:trivial] (a:bool) (b:bool) = a /\ not b
end

module Char
  (* utf8 characters (not glyphs)
      highly restricted until Why3 supports Unicode strings
  *)
  use int.Int
  use bv.BV256 as BV256

  type t

  function to_int (x : t) : int
  axiom to_int: forall c. 0 <= to_int c < 0x10FFFF /\ not(0xD800 <= to_int c < 0xE000)
  axiom extensionality:
    forall x y: t [to_int x, to_int y]. to_int x = to_int y -> x = y

  function of_int (n: int) : t
  axiom to_of_int: forall n [of_int n]. 0 <= n < 0x10FFFF /\ not(0xD800 <= n < 0xE000) -> to_int (of_int n) = n

  predicate le [@inline:trivial] (a:t) (b:t) = to_int a <= to_int b
  predicate lt [@inline:trivial] (a:t) (b:t) = to_int a < to_int b
  predicate ge [@inline:trivial] (a:t) (b:t) = to_int a >= to_int b
  predicate gt [@inline:trivial] (a:t) (b:t) = to_int a > to_int b

  let of_int (n : int) { 0 <= n < 0x10FFFF } { not(0xD800 <= n < 0xE000) }
    (ret (result : t) { to_int result = n }) = any

  let of_BV256 (n: BV256.t)
    { [@expl:arithmetic overflow] BV256.t'int n >= 0 /\ BV256.t'int n < 0x10FFFF }
    { not(0xD800 <= BV256.t'int n < 0xE000) }
    (ret (result: t)  { to_int result = BV256.t'int n}) = any
end

module Opaque
  type ptr
  let fresh_ptr (ret (result : ptr) {true}) = any

  type dyn
  type foreign
end
