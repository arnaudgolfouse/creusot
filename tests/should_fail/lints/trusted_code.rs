#![deny(creusot::trusted_code)]

extern crate creusot_contracts;
use creusot_contracts::prelude::*;

#[trusted]
#[ensures(result == 1i32)]
fn foo() -> i32 {
    1i32
}

#[trusted]
#[logic]
#[ensures(result == (x == 2i32))]
pub fn is_two(x: i32) -> bool {
    x == 2i32
}

#[allow(creusot::trusted_code)]
#[trusted] // Should not get flagged
#[ensures(is_two(result))]
pub fn bar() -> i32 {
    foo() + 1
}
