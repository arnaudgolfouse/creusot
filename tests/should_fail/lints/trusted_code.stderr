error: used the `#[trusted]` attribute
 --> trusted_code.rs:6:1
  |
6 | #[trusted]
  | ^^^^^^^^^^
  |
note: the lint level is defined here
 --> trusted_code.rs:1:9
  |
1 | #![deny(creusot::trusted_code)]
  |         ^^^^^^^^^^^^^^^^^^^^^
  = note: this error originates in the attribute macro `trusted` (in Nightly builds, run with -Z macro-backtrace for more info)

error: used the `#[trusted]` attribute
  --> trusted_code.rs:12:1
   |
12 | #[trusted]
   | ^^^^^^^^^^
   |
   = note: this error originates in the attribute macro `trusted` (in Nightly builds, run with -Z macro-backtrace for more info)

error: aborting due to 2 previous errors

